### PR TITLE
Google overlay state synchronization

### PR DESCRIPTION
**Pull request by felixpalmer**
https://github.com/visgl/deck.gl/issues/5980

#### Background

We were seeing issues with the integration where the Google basemap didn't display due to deck.gl changing WebGL state when the Google library didn't expect it

#### Change List

- Disable rendering using the deck.gl animation loop to ensure we only modify GL state in the Google `onDraw` lifecycle method
- Workaround for incorrect view state (hopefully we can remove this in the future if Google resolve on their end)
- Support rendering to framebuffer when present (supports optimized render path for Google maps)
- No longer need to continuously render (due to framebuffer fix)
- Animated layers and other state changed correctly request rerender via Google maps

### CODE REVIEW

**Review by Pessimistress — COMMENTED** [review]
What state exactly is being modified by deck and not reset other than the target framebuffer and viewport? Isn't  deck's viewport always set to the full canvas? If you search for `setParameters` in core there are not many places that it is used.

File: modules/google-maps/src/google-maps-overlay.js [hunk 1]
```diff
@@ -156,25 +156,39 @@ export default class GoogleMapsOverlay {
...
+      // As an optimization, some renders are to an separate framebuffer
+      // which we need to pass onto deck
+      const _framebuffer = getParameters(gl, GL.FRAMEBUFFER_BINDING);
+      deck.setProps({_framebuffer});
+
+      // Workaround for bug in Google maps where viewport state is wrong
+      // TODO remove once fixed
+      const viewport = getParameters(gl, GL.VIEWPORT);
+      const canvasViewport = [0, 0, gl.canvas.width, gl.canvas.height];
+      if (!canvasViewport.every((v, i) => v === viewport[i])) {
```
[the remark this answers is not in the document]
Comment (felixpalmer) [reply modules/google-maps/src/google-maps-overlay.js:179]: Done

File: modules/google-maps/src/google-maps-overlay.js [hunk 2]
```diff
@@ -156,25 +156,39 @@ export default class GoogleMapsOverlay {
     const deck = this._deck;
 
     deck.setProps({
-      ...getViewPropsFromCoordinateTransformer(this._map, coordinateTransformer)
+      ...getViewPropsFromCoordinateTransformer(this._map, coordinateTransformer),
+      _customRender: reason => {
+        if (reason !== 'viewState changed') {
```
[the remark this answers is not in the document]
Comment (felixpalmer) [reply modules/google-maps/src/google-maps-overlay.js:161]: Done

**Comment by felixpalmer** [comment]
@Pessimistress The viewport setting is for the benefit of Google Maps. Without that patch the rendering of the Google Map is wrong (clipped to [0, 0, 1024, 1024]), even if deck.gl works

Screen Shot 2021-09-08 at 2 23 56 PM

**Comment by felixpalmer** [comment]
Otherwise I don't think there is any state being modified, I've tried to make sure the `setParameters` call captures all that is necessary

**Review by alasarr — APPROVED** [review]
Thanks mate! Just tested in my local with some examples and It really fixes a bunch of issues.

I cannot reproduce the loading issue and it was happening to me half of the times

**Review by Pessimistress — APPROVED** [review]

**Pull request merged**